src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs

```diff
@@ -0,0 +1,353 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Resources;
+
+namespace OpenTelemetry.Exporter.Geneva;
+
+internal class OtlpProtobufSerializer
+{
+    private const int LengthAndTagSize = 4;
+
+    private readonly Dictionary<string, List<Metric>> scopeMetrics = new();
+
+    private int resourceMetricStartIndex;
+
+    private int scopeMetricsStartIndex;
+
+    private int metricStartIndex;
+
+    private int intrumentStartIndex;
+
+    private int metricPointStartIndex;
+
+    internal IMetricDataTransport MetricDataTransport;
+
+    public OtlpProtobufSerializer(IMetricDataTransport metricDataTransport)
+    {
+        this.MetricDataTransport = metricDataTransport;
+    }
+
+    internal void SerializeAndSendMetrics(byte[] buffer, Resource resource, in Batch<Metric> metricBatch)
+    {
+        foreach (var metric in metricBatch)
+        {
+            if (this.scopeMetrics.TryGetValue(metric.MeterName, out var metricList))
+            {
+                metricList.Add(metric);
+            }
+            else
+            {
+                var newList = new List<Metric>();
+                newList.Add(metric);
+                this.scopeMetrics[metric.MeterName] = newList;
+            }
+        }
+
+        // Serialize
+        this.SerializeResourceMetrics(buffer, resource);
+
+        this.ClearScopeMetrics();
+    }
+
+    internal void ClearScopeMetrics()
+    {
+        foreach (var list in this.scopeMetrics)
+        {
+            list.Value.Clear();
+        }
+    }
+
+    internal void SerializeResourceMetrics(byte[] buffer, Resource resource)
+    {
+        int currentPosition = 0;
+
+        this.resourceMetricStartIndex = currentPosition;
+
+        currentPosition += LengthAndTagSize;
+
+        // Serialize Resource
+        // TODO: Avoid serializing it multiple times.
+        SerializeResource(buffer, ref currentPosition, resource);
+
+        // TODO: Serialize schema_url field
+
+        // Serialize ScopeMetrics field
+        this.scopeMetricsStartIndex = currentPosition;
+        foreach (KeyValuePair<string, List<Metric>> metric in this.scopeMetrics)
```

Comment:
`metric` is not the best name for current element in the enumeration. Also, rename the dictionary `scopeMetrics` to something like `scopeToMetricsMapping` 

```suggestion
        foreach (KeyValuePair<string, List<Metric>> entry in this.scopeToMetricsMapping)
```